### PR TITLE
Ingest NPC AI XMLs

### PR DESCRIPTION
**Pull request by mettaursp**
added ingesting ai xmls & ai script names on npc ais
started attaching npc ai scripts to npcs (these ai scripts wont do anything yet, this is only data. need to implement behavior still)

### CODE REVIEW

File: Maple2.Server.Game/Manager/Field/AiManager.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+public sealed class AiManager {
+    private readonly ILogger logger = Log.Logger.ForContext<AiManager>();
+    public FieldManager Field { get; private set; }
+
+    public AiManager(FieldManager field) {
+        Field = field;
+    }
+
+
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Manager/Field/AiManager.cs:21]: empty lines

**Review by AngeloTadeucci — APPROVED** [review]

File: Maple2.File.Ingest/Mapper/AiMapper.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                return new AiMetadata.TrueCondition(node.name, childNodes.ToArray(), childAiPresets.ToArray());
+            default:
+                throw new NotImplementedException("unknown AI condition name: " + node.name);
+        }
+    }
+
+    AiMetadata.Node MapNode(Node node) {
+        List<AiMetadata.Node> childNodes = new List<AiMetadata.Node>();
+        List<AiMetadata.Condition> childConditions = new List<AiMetadata.Condition>();
+        List < AiMetadata.AiPreset> childAiPresets = new List<AiMetadata.AiPreset>();
```
Comment (AngeloTadeucci) [inline Maple2.File.Ingest/Mapper/AiMapper.cs:101]: formatting

File: Maple2.File.Ingest/Mapper/NpcMapper.cs [hunk 3]
```diff
@@ -23,6 +23,7 @@ public class NpcMapper : TypeMapper<NpcMetadata> {
             yield return new NpcMetadata(Id: id,
                 Name: name,
                 Model: data.model.kfm,
+                data.aiInfo.path,
```
Comment (Zintixx) [inline Maple2.File.Ingest/Mapper/NpcMapper.cs:26]: `AiPath: data.aiInfo.path,`

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 4]
```diff
@@ -210,6 +215,15 @@ public class FieldNpc : Actor<Npc> {
             foreach (string tag in Value.Metadata.Basic.MainTags) {
                 player.Session.ConditionUpdate(ConditionType.npc_race, codeString: tag);
             }
+		}
+	}
+	
+    public void SetAi(string name) {
+        AiMetadata? metadata = null;
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:222]: AiMetadata is already null, no need for this line

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 5]
```diff
@@ -80,6 +81,8 @@ public class FieldNpc : Actor<Npc> {
 
     public override Stats Stats { get; }
     public int TargetId = 0;
+    public string AiName { get; private set; }
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:84]: Is this needed? We can just get it from AiMetadata instead. Seeing as that's nullable and AiName is not may, it not be safe.

File: Maple2.Model/Metadata/AiMetadata.cs [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+
+public enum NodeBuffType {
+    remove,
+    add
+}
+
+public enum NodePopupType {
+    talk,
+    cutin
+}
```
Comment (Zintixx) [inline Maple2.Model/Metadata/AiMetadata.cs:71]: Move these enums to `Maple2.Model/Enum` 
Perhaps to a Ai.cs

**Review by Zintixx — CHANGES_REQUESTED** [review]

File: Maple2.Model/Metadata/AiMetadata.cs (hunk 6 above)
Comment (kOchirasu) [reply Maple2.Model/Metadata/AiMetadata.cs:71]: We should also use proper casing here

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 5 above)
Comment (mettaursp) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:84]: The AI script can be changed, and that is used for pets to switch between combat mode and out of combat. I'm not sure of any specific cases outside that yet, but it would also be good for debugging purposes to be able to set it on the fly.
`<v key="PetBattleAiPath" value="Pet/AI_DefaultPetBattle.xml" />`
AI_DefaultPetTaming is the one that is used for pets out of combat and completely lacks any ability to fight, while AI_DefaultPetBattle is used exclusively for combat.

Comment (kOchirasu) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:84]: But it's still redundant because if you are changing `AiName` here, you are also changing `AiMetadata` and `AiMetadata` already contains `Name`

Comment (mettaursp) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:84]: Ah now I see. I'll change it to a getter only using AiMetadata then

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 7]
```diff
@@ -210,6 +215,14 @@ public class FieldNpc : Actor<Npc> {
...
             }
+		}
+	}
+	
+    public void SetAi(string name) {
+        AiMetadata? metadata;
+
+        if (Field.AiMetadata.TryGet(name, out metadata)) {
+            AiMetadata = metadata;
         }
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:226]: this works?
```suggestion
        Field.AiMetadata.TryGet(name, out AiMetadata)
```

File: Maple2.File.Ingest/Mapper/AiMapper.cs [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+            childNodes.Add(MapNode(child));
+        }
+
+        foreach (AiPreset child in node.aiPreset) {
+            childAiPresets.Add(new AiMetadata.AiPreset(child.name));
+        }
+
+        switch(node.name) {
+            case "distanceOver":
+                return new AiMetadata.DistanceOverCondition(node.name, childNodes.ToArray(), childAiPresets.ToArray(), node.value);
```
Comment (kOchirasu) [inline Maple2.File.Ingest/Mapper/AiMapper.cs:68]: For consistency, we intentionally name all parameters when parsing. This will ensure fields continue to map properly in the event that types change.

For example: https://github.com/AngeloTadeucci/Maple2/blob/master/Maple2.File.Ingest/Mapper/ItemMapper.cs#L157-L182

File: Maple2.Model/Enum/NpcAi.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+public enum NodeTargetType {
+    Rand,
+    Near,
+    Far,
+    Mid,
+    NearAssociated,
+    RankAssociate,
+    Hasadditional,
```
Comment (kOchirasu) [inline Maple2.Model/Enum/NpcAi.cs:42]: HasAditional
RandAssociated
GrabbedUser

File: Maple2.Model/Enum/NpcAi.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
+﻿namespace Maple2.Model.Enum;
+
+public enum AiConditionTargetState {
+    GrabTarget,
+    Holdme
```
Comment (kOchirasu) [inline Maple2.Model/Enum/NpcAi.cs:5]: here and elsewhere, trailing comma on all lines in lists like these

File: Maple2.Server.Game/GameServer.cs [hunk 11]
```diff
@@ -25,6 +25,7 @@ public class GameServer : Server<GameSession> {
     private readonly ConcurrentDictionary<int, PremiumMarketItem> premiumMarketCache;
     private Dictionary<int, Shop> shopCache;
     private Dictionary<int, Dictionary<int, ShopItem>> shopItemCache;
+    public AiManager aiManager { get; set; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/GameServer.cs:28]: Do we really want `set` and not just `init`?

File: Maple2.Server.Game/Manager/Field/AiManager.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Maple2.Server.Game.Manager.Field;
+
+public sealed class AiManager {
+    private readonly ILogger logger = Log.Logger.ForContext<AiManager>();
+    public FieldManager Field { get; private set; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/Manager/Field/AiManager.cs:15]: Can we just use `init` if this is never going to change

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 13]
```diff
@@ -80,6 +81,8 @@ public class FieldNpc : Actor<Npc> {
 
     public override Stats Stats { get; }
     public int TargetId = 0;
+    public string AiName { get => AiMetadata?.Name ?? ""; }
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:84]: I'm not sure if I like this. I get why, as it's only a getter, but the Name for it isn't too much of a hassle to get, as it's right underneath `AiMetadata`

I'd suggest just removing this variable entirely and rely on getting it from AiMetadata.Name whenever you need.

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 14]
```diff
@@ -210,6 +215,12 @@ public class FieldNpc : Actor<Npc> {
...
                 player.Session.ConditionUpdate(ConditionType.npc_race, codeString: tag);
             }
-        }
+		}
+	}
+	
+    public void SetAi(string name) {
+        AiMetadata? metadata;
+
+        Field.AiMetadata.TryGet(name, out metadata);
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:224]: this func is currently not doing setting the metadata
```suggestion
        if (Field.AiMetadata.TryGet(name, out AIMetadata? metadata)) {
            AiMetadata = metadata;
        };
```

**Review by Zintixx — CHANGES_REQUESTED** [review]

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 7 above)
Comment (kOchirasu) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs]: This may accidentally make AiMetadata null

Comment (AngeloTadeucci) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs]: Isn't already null? what's the difference

Comment (kOchirasu) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs]: I assume this function exists to update the Ai of an existing Npc, so it may not be null and then become null

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 15]
```diff
@@ -210,6 +214,14 @@ public class FieldNpc : Actor<Npc> {
...
+
+        AiMetadata = AiMetadata;
```
Comment (kOchirasu) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:225]: this does nothing

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 7 above)
Comment (mettaursp) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs]: I can add a warning for when that happens, but npcs getting null ai metadata also isn't undesirable if there's a debugging command added for changing it. Pretty sure null metadata is something that should be expected since not all npcs even have the field set.
`11200003.xml: <aiInfo path="" />`

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 16]
```diff
@@ -210,6 +214,14 @@ public class FieldNpc : Actor<Npc> {
             foreach (string tag in Value.Metadata.Basic.MainTags) {
                 player.Session.ConditionUpdate(ConditionType.npc_race, codeString: tag);
             }
-        }
+		}
+	}
+	
+    public void SetAi(string name) {
```
Comment (kOchirasu) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:220]: you probably want a `[MemberNotNull(nameof(AiMetadata))]` here, might be a way to avoid marking `AiMetadata` as nullable

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 15 above)
Comment (mettaursp) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:225]: woops typo lol

File: Maple2.Server.Game/GameServer.cs [hunk 17]
```diff
@@ -25,6 +25,7 @@ public class GameServer : Server<GameSession> {
     private readonly ConcurrentDictionary<int, PremiumMarketItem> premiumMarketCache;
     private Dictionary<int, Shop> shopCache;
     private Dictionary<int, Dictionary<int, ShopItem>> shopItemCache;
+    public AiManager aiManager { get; init; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/GameServer.cs:28]: public properties should be capitalized, also why is this needed?

File: Maple2.Server.Game/Manager/Field/FieldManager.Factory.cs [hunk 18]
```diff
@@ -14,6 +14,7 @@ public sealed class Factory : IDisposable {
         // ReSharper disable MemberCanBePrivate.Global
         public required MapMetadataStorage MapMetadata { private get; init; }
         public required MapEntityStorage MapEntities { private get; init; }
+        public required AiMetadataStorage NpcAi { private get; init; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/Manager/Field/FieldManager.Factory.cs:17]: This should be called `AiMetadata` for consistency

File: Maple2.Server.Game/Manager/Field/FieldManager.cs [hunk 19]
```diff
@@ -60,6 +61,7 @@ public sealed partial class FieldManager : IDisposable {
     public int MapId => Metadata.Id;
     public readonly long OwnerId;
     public readonly int InstanceId;
+    public AiManager Ai { get; private set; }
```
Comment (kOchirasu) [inline Maple2.Server.Game/Manager/Field/FieldManager.cs:64]: why can't this be `public readonly AiManager Ai`?

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 16 above)
Comment (mettaursp) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:220]: Not all npcs have ai scripts on em. Nullable is required for AiMetadata
`11200003.xml: <aiInfo path="" />`

File: Maple2.Server.Game/GameServer.cs (hunk 17 above)
Comment (mettaursp) [reply Maple2.Server.Game/GameServer.cs:28]: forgot to remove lol

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 16 above)
Comment (kOchirasu) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:220]: Makes sense, what about `MemberNotNullWhen` instead?

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs (hunk 7 above)
Comment (kOchirasu) [reply Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs]: I think you should use a special case for that (e.g. empty string is passed in) rather than allowing it to be nulled whenever a script cannot be found.

File: Maple2.Server.Game/Manager/Field/FieldManager.Factory.cs (hunk 18 above)
Comment (kOchirasu) [reply Maple2.Server.Game/Manager/Field/FieldManager.Factory.cs:17]: Actually, is this even used? `AiMetadata` is already in `FieldManager`

**Review by Zintixx — APPROVED** [review]
